File: lib/models/misa.dart
enum MisaMode { general, farmer }

extension MisaModeX on MisaMode {
  String get label => this == MisaMode.general ? 'General' : 'Farmer-Specific';
}

enum MisaMessageAuthor { assistant, agent }

class MisaRecommendation {
  const MisaRecommendation({
    required this.title,
    required this.message,
    required this.actionLabel,
    required this.actionRoute,
    this.farmerId,
  });

  final String title;
  final String message;
  final String actionLabel;
  final String actionRoute;
  final String? farmerId;

  Map<String, dynamic> toJson() => {
        'title': title,
        'message': message,
        'actionLabel': actionLabel,
        'actionRoute': actionRoute,
        'farmerId': farmerId,
      };

  factory MisaRecommendation.fromJson(Map<String, dynamic> json) {
    return MisaRecommendation(
      title: (json['title'] as String? ?? '').trim(),
      message: (json['message'] as String? ?? '').trim(),
      actionLabel: (json['actionLabel'] as String? ?? '').trim(),
      actionRoute: (json['actionRoute'] as String? ?? '').trim(),
      farmerId: (json['farmerId'] as String?)?.trim(),
    );
  }
}

class MisaMessage {
  const MisaMessage({
    required this.id,
    required this.author,
    required this.message,
    required this.timestamp,
    this.recommendation,
  });

  final String id;
  final MisaMessageAuthor author;
  final String message;
  final DateTime timestamp;
  final MisaRecommendation? recommendation;

  Map<String, dynamic> toJson() => {
        'id': id,
        'author': author.name,
        'message': message,
        'timestamp': timestamp.toIso8601String(),
        'recommendation': recommendation?.toJson(),
      };
}

class MisaActionCandidate {
  const MisaActionCandidate({
    required this.id,
    required this.title,
    required this.summary,
    required this.actionLabel,
    required this.actionRoute,
    this.farmerId,
  });

  final String id;
  final String title;
  final String summary;
  final String actionLabel;
  final String actionRoute;
  final String? farmerId;

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'summary': summary,
        'actionLabel': actionLabel,
        'actionRoute': actionRoute,
        'farmerId': farmerId,
      };
}

class MisaRequest {
  const MisaRequest({
    required this.prompt,
    required this.mode,
    this.selectedFarmerId,
    required this.conversation,
    required this.context,
    required this.candidateActions,
  });

  final String prompt;
  final MisaMode mode;
  final String? selectedFarmerId;
  final List<MisaMessage> conversation;
  final Map<String, dynamic> context;
  final List<MisaActionCandidate> candidateActions;

  Map<String, dynamic> toJson() => {
        'prompt': prompt,
        'mode': mode.name,
        'selectedFarmerId': selectedFarmerId,
        'conversation':
            conversation.map((message) => message.toJson()).toList(),
        'context': context,
        'candidateActions':
            candidateActions.map((action) => action.toJson()).toList(),
      };
}

class MisaAiReply {
  const MisaAiReply({
    required this.message,
    this.recommendation,
  });

  final String message;
  final MisaRecommendation? recommendation;

  factory MisaAiReply.fromJson(Map<String, dynamic> json) {
    final recommendationJson = json['recommendation'];
    return MisaAiReply(
      message: (json['message'] as String? ?? '').trim(),
      recommendation: recommendationJson is Map<String, dynamic>
          ? MisaRecommendation.fromJson(recommendationJson)
          : null,
    );
  }
}

File: functions/api/misa.js
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
  'Cache-Control': 'no-store',
};

const model = '@cf/meta/llama-3.1-8b-instruct';

export async function onRequestOptions() {
  return new Response(null, {
    status: 204,
    headers: corsHeaders,
  });
}

export async function onRequestPost(context) {
  try {
    if (!context.env.AI) {
      return json(
        { error: 'Workers AI binding is not configured for this deployment.' },
        503,
      );
    }

    const body = await context.request.json();
    const prompt = typeof body?.prompt === 'string' ? body.prompt.trim() : '';
    const candidateActions = Array.isArray(body?.candidateActions)
        ? body.candidateActions.filter(isCandidateAction)
        : [];

    if (!prompt) {
      return json({ error: 'Prompt is required.' }, 400);
    }

    const systemPrompt = [
      'You are MISA, a field operations assistant for an agricultural workflow app.',
      'Use only the supplied context. Do not invent farmer state, actions, or routes.',
      'Your job is to answer the agent clearly and, when appropriate, choose exactly one allowed action from the candidate action list.',
      'Prefer short operational guidance over general advice.',
      'If the user asks about one farmer, keep the answer focused on that farmer unless the context clearly calls for escalation.',
      'Return ONLY valid JSON with this shape:',
      '{"answer":"string","recommended_action_id":"string or null","recommendation_title":"string or null","recommendation_message":"string or null"}',
      'If no listed action fits, set recommended_action_id, recommendation_title, and recommendation_message to null.',
      'Do not wrap the JSON in markdown fences.',
    ].join(' ');

    const transcript = Array.isArray(body?.conversation)
        ? body.conversation
            .filter((item) => item && typeof item.message === 'string')
            .map((item) => {
              const speaker = item.author === 'agent' ? 'Agent' : 'MISA';
              return `${speaker}: ${item.message}`;
            })
            .join('\n')
        : '';

    const userPrompt = [
      `Current agent request: ${prompt}`,
      transcript ? `Conversation so far:\n${transcript}` : null,
      `Mode: ${body?.mode ?? 'general'}`,
      `Selected farmer: ${body?.selectedFarmerId ?? 'none'}`,
      `Structured workflow context:\n${JSON.stringify(body?.context ?? {}, null, 2)}`,
      `Candidate actions:\n${JSON.stringify(candidateActions, null, 2)}`,
      'Remember: choose only from candidate_actions.id values when recommending an action.',
    ]
        .filter(Boolean)
        .join('\n\n');

    const result = await context.env.AI.run(model, {
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userPrompt },
      ],
      max_tokens: 450,
      temperature: 0.2,
    });

    const rawText =
        typeof result?.response === 'string'
            ? result.response
            : JSON.stringify(result ?? {});
    const parsed = parseModelJson(rawText);

    const answer = typeof parsed.answer === 'string' ? parsed.answer.trim() : '';
    if (!answer) {
      throw new Error('Model returned an empty answer.');
    }

    const actionId =
        typeof parsed.recommended_action_id === 'string'
            ? parsed.recommended_action_id.trim()
            : null;
    const chosenAction =
        actionId == null
            ? null
            : candidateActions.find((candidate) => candidate.id === actionId) ??
              null;

    const recommendation =
        chosenAction == null
            ? null
            : {
                title:
                    cleanOptionalString(parsed.recommendation_title) ??
                    chosenAction.title,
                message:
                    cleanOptionalString(parsed.recommendation_message) ??
                    chosenAction.summary,
                actionLabel: chosenAction.actionLabel,
                actionRoute: chosenAction.actionRoute,
                farmerId:
                    cleanOptionalString(chosenAction.farmerId) ??
                    cleanOptionalString(body?.selectedFarmerId),
              };

    return json(
      {
        message: answer,
        recommendation,
      },
      200,
    );
  } catch (error) {
    return json(
      {
        error:
            error instanceof Error
                ? error.message
                : 'MISA request failed.',
      },
      502,
    );
  }
}

function isCandidateAction(value) {
  return (
    value &&
    typeof value.id === 'string' &&
    typeof value.title === 'string' &&
    typeof value.summary === 'string' &&
    typeof value.actionLabel === 'string' &&
    typeof value.actionRoute === 'string'
  );
}

function cleanOptionalString(value) {
  return typeof value === 'string' && value.trim() ? value.trim() : null;
}

function parseModelJson(text) {
  const fencedMatch = text.match(/```(?:json)?\s*([\s\S]*?)```/i);
  const source = fencedMatch ? fencedMatch[1] : text;
  const start = source.indexOf('{');
  const end = source.lastIndexOf('}');

  if (start === -1 || end === -1 || end < start) {
    throw new Error(`Unable to parse model response: ${text}`);
  }

  return JSON.parse(source.slice(start, end + 1));
}

function json(payload, status) {
  return new Response(JSON.stringify(payload), {
    status,
    headers: {
      ...corsHeaders,
      'Content-Type': 'application/json',
    },
  });
}
